proposals/csharp-10.0/pattern-match-span-of-char-on-string.md

```diff
@@ -0,0 +1,49 @@
+<!--
+New language feature proposals should fully fill out this template. This should include a complete detailed design, which describes the syntax of the feature, what that syntax means, and how it affects current parts of the spec. Please make sure to point out specific spec sections that need to be updated for this feature. If you do not have the knowledge or experience to fill this out completely, that's perfectly alright: please open this proposal as a Discussion instead (https://github.com/dotnet/csharplang/discussions/new) and the community can generally discuss the proposal in less formal terms.
+-->
+# Pattern match `Span<char>` on a constant string
+
+* [x] Proposed
+* [x] Prototype: Completed
+* [x] Implementation: In Progress
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Permit pattern matching a `Span<char>` and a `ReadOnlySpan<char>` on a constant string.
+
+## Motivation
+[motivation]: #motivation
+
+For perfomance, usage of `Span<char>` and `ReadOnlySpan<char>` is preferred over string in many scenarios. The framework has added many new APIs to allow you to use `ReadOnlySpan<char>` in place of a `string`.
+
+A common operation on strings is to use a switch to test if it is a particular value, and the compiler optimizes such a switch. However there is currently no way to do the same on a `ReadOnlySpan<char>` efficiently, other than implementing the switch and the optimization manually.
+
+In order to encourage adoption of `ReadOnlySpan<char>` we allow pattern matching a `ReadOnlySpan<char>`, on a constant `string`, and optimize such switches on `ReadOnlySpan<char>` in the same way as switches on `string`.
```

Comment:
I would want to leave this up to the compiler, as there's definitely room for improvement with the current optimizations (in some cases it can be more efficient to emit a trie for example).

Reply:
Yes, but I don't think that detail needs to be in the specification itself, or the motivation. I think it's fine to just have the first part of the sentence and leave out the optimize part.

Reply:
@YairHalberstadt I'm just waiting on this before signing off.

Reply:
@jaredpar seemed to want more info on the optimizations, so was waiting for a conclusion :-)

Reply:
We don't have any info on optimizations, and I'd prefer to leave them out of the spec entirely.

Reply:
Change made